### ISSUE TITLE
consider the documentation of named event targets

### ISSUE THREAD

**Issue by AdrianRossouw** [nickfargo/state#15]
i've spent the last 2 weeks immersed in multiple systems using the state library, and only yesterday did I finally realise how the named events work.
### Setup

```
class LightSwitch
    state @,
        on: state 'initial',
          toggle: 'off'
        off:
          toggle: 'on'

light = new LightSwitch()        
```
### my naive understanding

even after having read that documentation several times, was that setting up a substate as a string, would cause that to to be the target when I transitioned to that state. and it would also fire the event (i know, i know).

```
# what i expected
light.state().path()     # 'on'
light.state('-> toggle') # 'off'
light.state('-> toggle') # 'on'
```

Allowing me to map the transitions between the states. I ended up hacking together something to do what I wanted.
### how it actually works

it wasn't until yesterday that I found out the following:

```
# what it actually works like
light.state().path()                  # 'on'
light.state().emit('toggle')        # null
light.state().path()                  # 'off'
light.state().emit('toggle')        # null
light.state().path()                  # 'on'
```

Once I knew that, I was able to just make a 'toggle' method to emit that, but the actual knowledge is pretty obliquely referred to in the documentation. I actually think there might be something broken in there ( but i am going to put that in another issue ).
### Why I intuited differently

The reason I found this incredibly counter-intuitive, is that in general I don't think it is an accepted pattern to use an event emitter to trigger changes on oneself. Like in backbone, I wouldn't consider it sound to add models to a collection by triggering the 'add' event.

Events _feel_ like they should be a notification of changes from within the object. There is of course the events-related command pattern, but it feels like that should be a separate emitter / verb
### How can the documentation be improved

There's no denying that the documentation is really technically dense, which is purely because of the completeness of the implementation.. but we might want to explicitly state this a bit more clearly.

The only place that the usage of strings as substates/named event targets is specifically mentioned in in [point 3 of interpreting session expressions], and the
only example that actually uses them is a [simple deterministic state machine](http://statejs.org/docs/#concepts--events--expressing-determinism), which is not really the friendliest example for this functionality.

**Comment by AdrianRossouw** [comment]
as an example here is a code pen of the above 'simple' example of the kid with his ice-cream, that I have adapted without the need to understand state bound functions and epistates

http://codepen.io/anon/pen/BKIHn

**Comment by nickfargo** [comment]
I have been leaning toward regarding this event-string-shorthand “feature” — i.e., interpreting the string-value of an event-key as an implicit transition expression — as having been a mistake, either in its present implementation, or likely its conceit altogether.

While it is concise, it’s also arbitrary, esoteric, presumptuous, unintuitive, and unfortunately as it turns out, misleading: perhaps most problematic is that it’s suggestive of an anti-pattern, namely, tightly entangling _observation_ with state _changes_. For simple cases such entanglement might be tolerable, but for anything non-trivial it can frequently lead to trouble (e.g., what if multiple listeners, in no particular defined order, each try to respond to a state event by instigating another transition?). In any case, it certainly seems like a pattern that should be used sparingly, if at all, and providing a shorthand for this I think may rightly be viewed as misguided.

So presently my thinking is that:
1. the general preferred syntax for expressing state events should be simply to provide a function expression (or array of functions) as the event handler(s), as one would with any familiar evented library;
2. code that effects changes to an object’s own state should generally be expressed as methods, while event handlers should generally be limited to expressing the observations of outside entities;
3. the [“Custom event types”](http://statejs.org/docs/#concepts--events--custom) example and [“Using events to express simple determinism”](http://statejs.org/docs/#concepts--events--expressing-determinism) section in the [docs](http://statejs.org/docs/) should each be deprecated, and either rewritten or removed.

Into the specifics …

---

##### To the `LightSwitch` example …

The points

> I don't think it is an accepted pattern to use an event emitter to trigger changes on oneself.

and

> Events _feel_ like they should be a notification of changes from within the object.

are well taken. What I would suggest in this example, then, is that `toggle` be implemented as a method, rather than an event:

``` coffee
class LightSwitch
  state @::,  # <-- i.e. `LightSwitch.prototype`, not the `LightSwitch` constructor
    on: state 'initial',
      toggle: -> @state '-> off'
    off:
      toggle: -> @state '-> on'
```

… which of course is expressed here with considerable sugar and shorthand, and will be interpreted as equivalent to:

``` coffee
class LightSwitch
  state @::,
    substates:
      on: state 'initial',
        methods:
          toggle: -> @state().go('off')
      off: state
        methods:
          toggle: -> @state().go('on')
```

This then should allow for user code that’s more idiomatic and, hopefully, intuitive:

``` coffee
light = new LightSwitch
light.state().path()  # >>> 'on'
do light.toggle
light.state().path()  # >>> 'off'
do light.toggle
light.state().path()  # >>> 'on'
```

Verbs that affect the `light`’s state are kept as methods, as they should be, and events, if you choose to add them, should generally limit themselves to observation from outside.

---

##### To the codepen example …

http://codepen.io/anon/pen/BKIHn

The fix I would apply is to change the `Kid` class as written from:

``` coffee
# had to use the constructor here
# because 'virtual epistates' don't inherit
# a bunch of things.
class Kid
  constructor: -> 
    state @,
      Happy: state
      Sad: state
      events:
        spilledIceCream: 'Sad'
        gotIceCream: 'Happy'
```

to the more imperative and verbose, but more explicit:

``` coffee
class Kid
  state @::,
    Happy: state
    Sad: state
    events:
      spilledIceCream: -> @state '-> Sad'
      gotIceCream: -> @state '-> Happy'
```

(or if you prefer/need to bind to the `State` object for whatever reason):

``` coffee
class Kid
  state @::,
    Happy: state
    Sad: state
    events:
      spilledIceCream: state.bind -> @go 'Sad'
      gotIceCream: state.bind -> @go 'Happy'
```

---

With regard to the note about virtual epistates not inheriting a bunch of things — if there are also problems related to anything in addition to this event-sugaring issue, I’d be very interested to hear more about anything you’ve found.

Great feedback, thanks.

**Comment by AdrianRossouw** [comment]
i find being able to specify target states in string form (versus requiring a method) ridiculously useful, and it is something i will be making use of on an almost daily basis. If it is removed I can probably end up making a closure generating utility function to replicate it however.

### use case

The use case for this is the most eloquently expressed in something as straight forward as a wizard, where you want to specify a 'next' link. or maybe a workflow state to go to onComplete or onFailure.

```
class Wizard
  # go to the next page, wherever it is
  next: ->   @state().emit('next')
  configure: ->
    state @,
      start:
        next: 'page1'
      page1:
        next: 'end'
      end: state 'final'
```

### background on my earlier issues

This is something I needed a lot for my node knockout competition entry, where I had to maintain a server side collection of backbone models that represented the player sessions, and it had to keep track of what level of connection the player had to us. so it would go offline -> session -> websocket -> sip -> voice.

Not only that, the state machine had to constantly be trying to upgrade and downgrade the connection status, so it would handle situations like only making the webrtc connection in one browser tab at a time, and when you closed it, it would automatically dial in on a second open tab.

This is the hack I eventually came up with to manage it ([gist](https://gist.github.com/Vertice/1ba7820723b571c4345a#file-session-coffee-L48-L57)), where I even had to re-implement the admit guard as 'allow', because I ran into so many unexpected behaviours with them.

### today and yesterday's adventure.

So finally figuring out the event handler things, I spent some time yesterday converting that code into what I hoped would be a much cleaner solution, and running them against the unit tests I already had in place, because I am going to make significant use of the feature in my day job.

What I came up with is quite elegant, and is marvellously expressive 

```
upgrade: ->
  before = @state().path()
  @state().emit 'upgrade'
  @upgrade() if before != @state().path()
```

Lovely as that was, try as I might, things were still just horribly breaking all the time, and having lost many hours to this over the last few weeks I resolved to work through all of the various combinations of things I have tried. I started making individual code pens for each of the different approaches I attempted, so I could pinpoint where the issue was.

The full [codepen example listing](http://www.one-tab.com/page/zWNxsWZ1TEm4XiI4WSZDLw)
1. [example 1](http://codepen.io/anon/pen/DhGox) is how I expected it to work. It would only make the first transition, and then just stop. In my debugger, the guards would get fired, but the return values seemed to be completely ignored.
2. [example 2](http://codepen.io/anon/pen/jpsbL) is when i stripped it down to the basics, to just get the basic functionality working.
3. examples 3 through 6 were me systematically going through each of the possible things that could resolve it, such as wrapping the admit functions in state.bind,and various other combinations of things.
4. [example 7](http://codepen.io/anon/pen/Khuzt) is where I got it working again, by using the object form of the admit callback (specifying the origin). I swear I had tried this multiple times, and there were several occasions where it would just fail for no reason.. but i'd spent enough time on this.
5. [example 8](http://codepen.io/anon/pen/wsjzd) is where I turned the key of the admit function, into a comma-separated string of states. This is something I only knew about because I had seen the key being split in the code while stepping through it.

This was also complicated by the fact I have to persist and inflate the sessions into a database, so the initialize function needs to be able to 'inflate' the state. Hence the admit guards would need to be fired from multiple places, and i really wanted to avoid creating a state bound function just to share the conditions across multiple origin paths.

**Comment by nickfargo** [comment]
Very interesting — great step-by-step feedback again.

The way you use the event-string form here is certainly elegant. I fear for more complicated use cases, but perhaps that syntax deserves to live on. I’ll certainly have to give that more thought before acting on anything.

---

Anyway, your experience leads me toward a couple conclusions.

First, the existing documentation in the [Guards](http://statejs.org/docs/#concepts--guards--state-guards) section needs to be improved. What’s there illustrates the point, but barely, and it’s pretty cold and algebraic looking.

Second, what you tried initially, i.e. a plain function/predicate

``` coffee
socket: state
  admit: -> true if @owner.foo
```

is completely intuitive — indeed I was stumped for a few minutes myself as to why that didn’t work — and I see no reason that that should not be accepted by the `StateExpression` interpreter, and understood as equivalent to the same predicate hung off a double-wildcard (“any state”) selector:

``` coffee
socket: state
  admit: '**': -> true if @owner.foo
```

As I think you tenaciously discovered, then, what the `admit` guard expression is expecting is a provided hash/object where a property’s _key_ identifies the state(s) to which the boolean or predicate in the corresponding _value_ applies:

``` coffee
socket: state
  admit:
    session: true
```

If the `admit` value is not an object, but rather a string, then one shorthand that is presently accepted by the interpreter is to assume an associated `true` predicate:

``` coffee
socket: state
  admit: 'session'
```

So for now, an expected form for the pattern you meant to express in the `Session` class could have been simply:

``` coffee
class Session
  constructor: ->
    @initState()

  initState: ->
    state @,
      upgrade: ->
        before = @state().path()
        @state().emit 'upgrade'
        log before, '--->', @state().path()

      offline: state 'initial',
        upgrade: 'session'

      online: state 'abstract',
        session: state
          upgrade: 'socket'
          admit: 'offline'
        socket:
          upgrade: 'sip'
          admit: 'session'
        sip:
          upgrade: 'voice'
          admit: 'socket'
        voice:
          downgrade: 'sip'
          admit: 'sip'
```

… no need to rely on the expandos (`ses.session`, etc.) to determine eligibility for `admit` (if that’s what they’re for).

I think we get a nicely terse, expressive form here … though, again, your description of the experience getting to this point was very helpful, as it seems that the way guards are expressed could stand to be more accommodating.

---

Meanwhile I think I’m also seeing now some of the issues when implementing `state` on the class’s prototype, where events are not inheriting via protostates/epistates when used in this way … we’ll see where this leads and report back …

**Comment by AdrianRossouw** [comment]
there's just something ... funky with how the guards are evaluated. if you only have the admit up to 'session' (ie: the first jump), it works fine ... but the further jumps fail strangely.

The flow of this code, from memory (i also can't link to line numbers in .md files...),

in the debugger you reach a point where:
- it evaluates the release guards and then the admit guards.
- the arguments are origin:target / target:origin (or vice versa)
- the key gets defaulted to '*' (this is close to where i saw the key being split)
- when doing the admit guard, it does a context.query(key, somethingOrOther)
- this is meant to determine whether your origin and your destination are the same/different..
- this causes the for loop to continue, instead of passing on to the function call to the guard.

When i was testing this in the console it seemed that context.query('**', somethingOrOther) would
indeed return the result that would allow it to pass further and fire things.

At the time i was still using the nested form (online.session), so i dont know if changing what the
default is will work, or what else it will break in the rest of the system.

And not just events weren't working when inherited that way, transitions were completely broken. I
was intending on building on those (ended up being cleaner without them), but there was a good few
hours where I was completely lost.

And those properties are actually session id's for the various communication channels relating to the
session, so i could do a lookup if i get an event from the voip side of things, and map that to
the relevant connect session and hence the relevant socket id.. etc.

the project: [werewolves.io](http://nodeknockout.com/teams/nodesque). I am still working on it though,
so let's just say I am pretty immersed in state.js at the moment.

**Comment by nickfargo** [comment]
> there's just something ... funky with how the guards are evaluated. if you only have the admit up to 'session' (ie: the first jump), it works fine ... but the further jumps fail strangely.

Hmm … well first, just to make sure we’re seeing the same thing … the final code block in my previous post [and in this pen](http://codepen.io/anon/pen/riqbw) is giving me what I understand to be the expected behavior:

``` coffee
class Session
  constructor: ->
    @initState()

  initState: ->
    state @,
      upgrade: ->
        before = @state().path()
        @state().emit 'upgrade'
        log before, '--->', @state().path()

      offline: state 'initial',
        upgrade: 'session'

      online: state 'abstract',
        session: state
          upgrade: 'socket'
          admit: 'offline'
        socket:
          upgrade: 'sip'
          admit: 'session'
        sip:
          upgrade: 'voice'
          admit: 'socket'
        voice:
          downgrade: 'sip'
          admit: 'sip'


ses = new Session()
log 'starts as', ses.state().path()
ses.upgrade()
ses.upgrade()
ses.upgrade()
ses.upgrade()
```

```
starts as offline
offline ---> online.session
online.session ---> online.socket
online.socket ---> online.sip
online.sip ---> online.voice
```

Do I have this right?

**Comment by AdrianRossouw** [comment]
yeah. that's right

like i said, it was from memory, and i was mentioning it because it was a specific case I recall the '*_'
would have made a difference to what i was seeing. I also seem to remember the key being a boxed
object, that even through the debugger only showed it as 'keyname', ended up having a '_': [function],
property.

but i'll leave you to figure out what is actually going on there =)

**Comment by nickfargo** [comment]
Yeah, seems it’d make sense if it were a matter of the `*`/`**` distinction — where `*` selects just the immediate substates of the root, `offline` and `online`, and excludes the nested substates of `online`, while `**` selects all descendants of the root.

(another thing that I myself forgot about for a minute earlier today)

In fact, come to think … it’s very possible that you’ve tracked down a bug in [`StateExpression interpret`](http://statejs.org/source/state-expression.html#state-expression--private--interpret), under the “Coercion” section, where it rewrites the guards:

> Guards are represented as an object keyed by selector, so non-object values are coerced into a single-element object with the value keyed to the wildcard selector `*`.

``` coffee
    for own key, value of object = result.guards
      object[ key ] = '*': value unless isPlainObject value
```

and we probably want that key to be a `'**'` instead.

I’ll give that a hard look and conjure up some additional tests … thanks for the eyes on this.

**Comment by AdrianRossouw** [comment]
more fun with these things today i'm afraid. see [example pen](http://codepen.io/anon/pen/Gvlos).

So it seems like named event targets will only work one level deep. unless of course abstract states get inherited. I'll report back with how i worked around this.

**Comment by nickfargo** [comment]
I think we might be ok this time … [pen … pong](http://codepen.io/anon/pen/cKkbG)

[`State::emit`](http://statejs.org/api/#state--methods--emit)’s `via` param defaults to the traversal flag `VIA_ALL`, therefore the fix is just

``` coffee
@state().emit 'next', VIA_NONE
```

With no `via` provided, and the bit-masked `VIA_SUPER` thus in effect … `emit` on `stateB` was emitting its own `next`, then inheriting `stateA`’s `next` immediately afterward … and the transition was redirected whence it came, trapped in `stateB`.

**Comment by AdrianRossouw** [comment]
that's not esoteric in the slightest =)

I ended up handling it via a closure in my state util library.

```
// returns a closure that will change to a new location.
function go(path, pushState) {
    pushState = _.isUndefined(pushState) ? true : pushState;

    return function() {
        var before = this.state().path();
        this.state().go(path);
        var changed = (before !== this.state().path());
        if (changed && pushState) {
            this.setUrl(this.state().path());
        }
    };
}
```

so i leave it in place like :

```
general: state({
    view: sutil.view('GeneralInfo'),
    next: sutil.go('languages'),
    back: sutil.go('..')
}),
```

**Comment by nickfargo** [comment]
> that's not esoteric in the slightest =)

Yeah, one of those things. I _think_ `VIA_ALL` is the proper default, but then, there’s a chance that it ought to be `VIA_PROTO`, or something, instead. Hard to intuit intuition.

I will say, though, that I believe the trouble we’re seeing here harkens right back to points alluded to in the first couple paragraphs at the top of this issue: that it’s hazardous, fundamentally, to entangle _observation_ in a causal relationship with _state changes_.

This is why the standing recommendation is, generally:
1. to express code that would affect an object’s own state in terms of a _method_, which is only ever dispatched to one location, and control is then only explicitly delegated elsewhere (e.g. a superstate) by the user before returning to the call site; and
2. to restrict _events_ — for which there may be any number of listeners, that respond to any and all comers, and that are bound to and inherited from any number of possible relatives in the state graph — as much as possible to code that only involves concerns outside the purview of the object, and can be expected not to compete for priority or attention from the object.

**Comment by nickfargo** [comment]
I much prefer the method-based approach of your util factory, by the way.

And, thinking about this some more — while I’m still not keen on this string-as-transition-effect-of-event-handler thing — the hazard of

``` coffee
state this,
  aSuperstate: state
    next: 'aSubstate'
    aSubstate: state 'initial'
      next: 'anotherState'
  anotherState: state

@emit 'next'
@state().is 'anotherState'  # >>> false (should be true)
```

could be mitigated by making sure that only the transition side-effect caused by the “least inherited” string-as-event-handler (here, `aSubstate:next`) would be executed.

(In the source: return the `eventualTarget` state from `StateEventEmitter::emit` back to `State::emit`, and let the inheritance recursion in `State::emit` filter only the least inherited one.)

These event side-effects still probably aren’t a good idea, but at least this would considerably lessen the confusion of those effects across inherited relations.

**Comment by nickfargo** [comment]
During the latest commit d6d0c052fe94f2456ec0b6b304fc92041948e2fd I realized [I was mistaken earlier](#issuecomment-29039307) about the shorthand syntax for guards, where:

``` coffee
      offline: state 'initial',
        upgrade: 'session'

      online: state 'abstract',
        session: state
          upgrade: 'socket'
          admit: 'offline'
```

does not work in v0.2.0, which expects

``` coffee
          admit: { 'offline': true }
```

So, transformation of bare string values as guard expressions in this manner was added in that commit. And, in that vein, next I’ll start experimenting further with rebuilding the guard structure as per the ideas in #17.

**Comment by AdrianRossouw** [comment]
The only reason I was using the 'emit' form was because it was clearly what the API was designed for and wanted me to do.

My gut feel to try removing most magic string parsing and providing closure constructing helper functions in the library that are clearly documented. `eventName: state.go('statename')`

**Comment by nickfargo** [comment]
I have the same general feeling as well, in that I dislike “magic” to the extent that it suggests unpredictability. Leaning on string-parsing patterns can get confusing fast, even when limited to very short strings, unless it’s planned and thought out exhaustively.

Then on the other hand, there may be something to be said for still having the option of reducing the contents of one’s state expressions down to primitives, or at least some kind of stateless entities, such as strings, when the situation calls for it (ease of JSON stringification comes to mind).

hmm …

**Comment by AdrianRossouw** [comment]
I really considered the json situation too. especially since I will need to dynamically generate state machines based on workflow diagrams built by users in a browser. I've also spent a lot of time working with [json-schema](http://json-schema.org) and various libraries.

You know what JSON doesn't support? at all? functions. Basically, the moment you have a function in the data structure, as part of the data structure. I don't think you can really express it as JSON anymore.

You can try to hack around that all you want with fn.toString(), but at that point it all just becomes convention. You really aren't using it as an interchange format anymore, and it's not really reasonable to expect the serialised functions to carry scope with them.

Another thing to note, and what I feel the the implementation really implies, is that you want to be able to cleanly and concisely serialize the OWNER in this relationship, not the state machine itself. It really feels like it's reason for existing in the first place.. attaching all that stuff that is inherently incompatible with all transit formats, to a persistable owner object.

What I also really really like about the closure method, is how much more testable it is. You can write some incredibly simple, yet incredibly tight unit tests using [spies and mocks](http://sinonjs.org) so that you know with absolute and utter certainty, no matter what else is happening in the definition around this call, it will always call the objects state().go function. always.

**Comment by AdrianRossouw** [comment]
oh. i should add one place where i feel that strings would be best used.

I feel that a bare string, should probably be the same as calling that method on the object, in the right scope.

```
class Session
  state @::
    online: state
      validate: -> !!@sessionId
      admit: 'validate'
    offline: state
      validate: -> !@sessionId
      admit: 'validate'
```

Or something similar.

**Comment by nickfargo** [comment]
> JSON … functions.

Yep, if we want persistence, we can’t have hidden lowercase-s state lurking about in a closure. That really is the main thing that keeps tugging me back toward using string patterns, to promote portability of state expressions — at least, to the extent that’s possible.

I’ve not yet studied or thought out the persistence angle very far, though; this ideal may well be a fruitless ideal, I don’t know.

> I feel that a bare string, should probably be the same as calling that method on the object, in the right scope.

Hm. Yeah, that is compelling, indeed.

But then, that leaves us with a collision between two compelling interpretations:

``` coffee
online: state
  admit: 'validate'  # delegate this guard’s predicate to the `validate` method
```

and

``` coffee
online: state
  admit: 'offline'  # admit the transition if coming from the `offline` state
```

and a dilemma. Hmm …

**Comment by AdrianRossouw** [comment]
Well, I'm not sure what anybody else's implementations look like, but I am using the state machine to swap out different methods attached to the owner to be dispatched, based on the state.

My dispatched methods are definitely not something I even remotely expect to be serializable, hence I don't really expect anybody else's actual business objects to be, regardless of what the guards/events look like. Not only that, but without those methods being persisted and inflated correctly, the entire thing is a moot because it really doesn't gain me anything other than a broken instance.

Regardless of whether it could be done in the first place, having a well defined set of verbs/actions as helper method actually gets you further to the goal of persistence because you can turn it into JSON via a convention quite similar to how you box functions for state.bind.

I think the first case is far more general and could probably be applied in a consistent manner to gain a whole range of functionality. The latter is a special case only applicable to guards. I'm really trying to think of ways to avoid as many as possible special cases.

**Comment by zoomclub** [comment]
Hi Nick,

Good to see all the progress on State.js. I've been plugging away here bringing multiple facets of a larger app into order, once cleaned up a bit more I'll be ready for State.js :)

Just saw JSON being mentioned and wanted to chime in and say that having a JSON based machine would be awesome. I'm using it as a DI solution throughout my app and storing JSON with Firebase. By activating JSON injects I can change the behaviour of my tool objects at different points of the app.

Here are two very good examples of where JSON is being used well:

http://trifacta.github.io/vega/

Checkout the Vega Live Editor

http://krasimir.github.io/absurd/

Scroll down on page for more

Cheers,

Thom

**Comment by nickfargo** [comment]
@Vertice 

> Regardless of whether it could be done in the first place, having a well defined set of verbs/actions as helper method actually gets you further to the goal of persistence because you can turn it into JSON via a convention quite similar to how you box functions for state.bind.
> 
> I think the first case is far more general and could probably be applied in a consistent manner to gain a whole range of functionality. The latter is a special case only applicable to guards. I'm really trying to think of ways to avoid as many as possible special cases.

@zoomclub 

> By activating JSON injects I can change the behaviour of my tool objects at different points of the app.

Lots of strong points. The bit about generality versus special cases is a particularly good one.

I had to think on this for a while … I ended up cobbling together most of those thoughts into the proposal at #19 … probably could use more than a tweak or two, but I expect that have a chance to lead to something a lot clearer, more regular and explicit, and perhaps, significantly more serializable.

**Comment by nickfargo** [comment]
@Vertice 

> My gut feel to try removing most magic string parsing and providing closure constructing helper functions in the library that are clearly documented. `eventName: state.go('statename')`

I have been turning this idea in particular over in my head for a while, and now must admit I’m coming up short on how this would help give us a favorable alternative to parsing strings magically in one form or another.

I can see `eventName: state.go('stateName')` as a way to carve out an alternative to the current behavior — i.e., of accepting a string as a transition target selector and assuming an imperative to execute a transition to that target state as a consequence of the event — in order to make room for the alternative of interpreting a bare string in that context as naming a method, or what have you.

I can also see this as a way to be more explicit about indicating intent, for this case — i.e., a bare state name provides an “objective” but specifies no “action” (namely, the transition). Likewise this reasoning might also suggest the propriety of a `state.delegate('methodName')`, rather than leaving the alternative’s intent implicit as well.

What I don’t see is how a static `state.go` factory (or the `bind` or `fix` boxing functions for that matter) would be helping to create opportunities to define state expressions that are themselves more stateless, and therefore serializable — which I think is the higher goal in all of this. Whether a function is constructed or boxed at runtime, it’s still a `function`, and therefore must be assumed to be “hiding something” (i.e., lowercase-s state).

So for the particularly common use cases — e.g., transitioning to a target state; or delegating to a named method — if there’s a way to reduce them to a simple, stateless primitive, such as a string, then I think that seems like the most favorable alternative.

This is why, for the moment, I find myself leaning toward “magic string parsing” for basic selectors and commands — to transform imperative into declarative — but only in such a way that demystifies the magic, by defining a simple grammar that would make the parsing rules explicit, and for us human-folk, make the semantics predictable and easy to grok.

I expect I’m still blind to something here, though — maybe you can shed a bit more light for me.
